python/phase-1.md

```diff
@@ -0,0 +1,380 @@
+exported from `temporalio`):
+
+```python
+class Worker:
+    def __init__(
+        self,
+        client: temporalio.Client,
+        *,
+        # Uses default in client if not present (or errors if no default)
+        task_queue: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activity(
+        # This must be a callable. Can be a lambda.
+        activity,
+        *,
+        # Set the name override. If not set, uses the @activity.run decorator
+        # name. If that's not set, uses the function/method name or class name
+        # if it has __call__. If it's a lambda or nested function, a name must
+        # be provided (or it can be decorated of course).
+        name: str = None
+        # ...
+    ) -> None:
+        pass
+
+    # Cannot be called after run started. Typings and separate @overload
+    # signatures omitted for brevity.
+    def register_activities(
+        # This must be an object or list or dictionary. If an object, this only
+        # uses methods with @activity.run decorators. Dictionaries and lists
+        # don't require the decorator. If a dictionary, the key is the name.
+        # This is just a helper and is the equivalent of just calling
+        # register_activity with all applicable functions.
+        activities,
+        # ...
+    ) -> None:
+        pass
+
+    async def run() -> None:
+        pass
+```
+
+Notes:
+
+* We have left off workflow registration until a later phase
+* Stopping the worker is done via cancellation of the run task/future
+* Internally, activities will run in the same event loop `run()` is called on
+* By default will fail already-registered activities
+* Should we accept a set of activities to register in `__init__`?
+* Does exporting `temporalio.Worker` directly cause any packaging issues if we want to separate the client as its own
+  published package?
+
+### Activity Definition/Registration
+
+```python
+from temporalio import activity
+
+# This adds activity details if they are present. Maybe we should have a
+# temporalio.LoggerAdapter that works for activities _and_ workflows?
+my_activity_logger = activity.LoggerAdapter(logging.getLogger(__name__))
+
+
+def simple_activity(arg1, arg2):
+    # Uses contextvars internally, which means it'll fail in another context but
+    # we will provide the ability to set this for testing situations
+    attempts = activity.info().attempts
+    # This logger has an adapter to include custom details but still named after
+    # this module which is a common pattern
+    my_activity_logger.debug("Attempt %s", attempts)
+    # This logger, on the activity module, uses the name of the activity module.
+    # So that's why we offer either the adapter or the logger.
+    activity.logger().debug("Attempt %s", attempts)
+
+
+@activity.run(name="MyActivity")
```

Comment:
Setting the activity name at the same spot as the function name is reasonable (it can be registered to lots of workers). The decorator also serves other purposes such as for people that want to mark their activities clearly (though obv not required) and its required for activity class registration to mark which methods are activities.

Reply:
One thing we can do with the decorator is set the default activity options.
These options can be used in workflows running in the same worker that the activities are registered in.

Reply:
Another options for activity class registration is to expose all public async functions by default.